docs/rfcs/confmap-provider-configuration.md

```diff
@@ -0,0 +1,246 @@
+without a URI and to ensure the options are always applied to all URI
+resolutions.
+
+### Configure options inside the URI
+
+[RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3), which
+specifies the format of a URI, specifies the different parts of a URI and
+suggests two places where we could pass options to Providers: queries and
+fragments.
+
+#### Queries
+
+Breaking changes:
+
+- confmap Providers would have breaking changes since they would now consume URI
+  queries. There would be no breaking changes to the confmap API.
+
+Advantages:
+
+- Explicitly intended to specify non-hierarchical data in a URI.
+- Often used for this purpose.
+- Fits into existing config URIs for URL-based Providers.
+
+Disadvantages:
+
+- Completely consuming the parameters would mean we can't pass them to config
+  backends. The frequent use of query parameters in server endpoints would force
+  us to limit the types of requests we can make.
+- Partially consuming parameters would still cause conflicts if a Provider
+  parameter and config backend parameter share the same name. Additionally,
+  mixing the two would provide suboptimal UX.
+- Only allows easily specifying key-value pairs.
+
+#### Fragments
+
+We could specify options in a query parameter-encoded string placed into the URI
+fragment.
+
+Breaking changes:
+
+- confmap Providers would have breaking changes since they would now consume
+  fragments. There would be no breaking changes to the confmap API.
+
+Advantages:
+
+- Not likely to be used by config backends for any of our supported protocols,
+  so has a low chance of conflict.
+- Fits into existing config URIs for URL-based Providers.
+
+Disadvantages:
+
+- Even if fragments are likely not useful to backends, we are still preventing
+  their use in upstream Providers.
+- Doesn't conform to the spirit of how fragments should be used according to RFC
+  3986.
+- Will not work with Providers like the env Provider. This is likely minor, as
+  the env Provider can be expected to work differently than URL-based Providers.
+- Only allows easily specifying key-value pairs.
+
+We could likely partially circumvent the key-value pair limitation by
+recursively calling confmap Providers to resolve files, env vars, HTTP URLs,
+etc. For example:
+
+```text
+https://config.com/config#refresh-interval=env:REFRESH_INTERVAL&headers=file:headers.yaml
+```
+
+Using this strategy would also allow us to more easily get env vars and to get
+values from files for things like API tokens.
+
+### Add to the string passed to the config flag
+
+Instead of configuring Providers inside their URIs, we could add an extra
+section to the string passed to `--config` and configure how a URI is retrieved
+through that config. This approach is undesirable because the config flag
+currently only takes valid URIs, and updating the format to accommodate this
+would require we adopt an unconventional format.
```

Comment:
The more I read, the more I would prefer something like this: no conflict, it's clear what it means, you can configure per URI... This seems easy enough in the config itself (e.g. we could have `${http://example.com}[refresh-interval=10s]` or something like this), but I don't see now how to put that into practice for the CLI right now :thinking:

Reply:
I would see it work like this for each format:

File:

```
val: ${http://example.com}[refresh-interval=10s]
```

CLI:

```
--config=http://example.com
--resolution-options=refresh-interval=10s&auth-header=Api-Token
```

If we don't do a config flag, we'll have to find some character that's not supported [here](https://www.rfc-editor.org/rfc/rfc3986#appendix-A) and use it as a delimiter between the URI and options. Looking at the `gen-delims` and `sub-delims` character classes, `|`, `\`, `{`, `}`, `^`, and `` ` `` could be options.